cpp/lib/ResultSetArrow.cpp

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright (c) 2021 Snowflake Computing, Inc. All rights reserved.
+ */
+
+#include <cmath>
+#include <iomanip>
+#include <memory>
+#include <sstream>
+#include <time.h>
+
+#include "../logger/SFLogger.hpp"
+#include "ResultSetArrow.hpp"
+#include "result_set_arrow.h"
+#include "results.h"
+
+
+namespace Snowflake
+{
+namespace Client
+{
+
+
+ResultSetArrow::ResultSetArrow() :
+    Snowflake::Client::ResultSet()
+{
+    m_queryResultFormat = QueryResultFormat::ARROW;
+    m_bufferBuilder = std::make_shared<arrow::BufferBuilder>();
+}
+
+ResultSetArrow::ResultSetArrow(
+    cJSON * initialChunk,
+    SF_COLUMN_DESC * metadata,
+    std::string tzString
+) :
+    ResultSet(metadata, tzString)
+{
+    m_queryResultFormat = QueryResultFormat::ARROW;
+    m_bufferBuilder = std::make_shared<arrow::BufferBuilder>();
+    init(initialChunk);
+}
+
+// Public methods ==================================================================================
+
+SF_STATUS STDCALL ResultSetArrow::appendChunk(cJSON * chunk)
+{
+    // Decode Base64-encoded Arrow-format rowset of the chunk.
+    const char * base64Rowset = snowflake_cJSON_GetStringValue(chunk);
+    if (base64Rowset == NULL)
+    {
+        CXX_LOG_ERROR("Unable to retrieve value of key 'base64Rowset'.");
+        return SF_STATUS_ERROR_BAD_RESPONSE;
+    }
+
+    std::string decodedRowset = arrow::util::base64_decode(std::string(base64Rowset));
+    m_bufferBuilder->Append((void *)decodedRowset.c_str(), decodedRowset.length());
+
+    CXX_LOG_TRACE("Chunk %d received.", m_currChunkIdx);
+    m_currChunkIdx++;
+
+    // Finalize the currently built buffer and initialize reader objects from it.
+    // This resets the buffer builder for future use.
+    std::shared_ptr<arrow::Buffer> arrowResultData;
+    m_bufferBuilder->Finish(&arrowResultData);
+    m_bufferReader = std::make_shared<arrow::io::BufferReader>(arrowResultData);
+    arrow::ipc::RecordBatchStreamReader::Open(m_bufferReader, &m_batchReader);
```

Comment:
arrow::ipc::RecordBatchStreamReader::Open(std::move(m_bufferReader)) 
Would be better.

Reply:
Great catch, thank you!